### PR TITLE
fix: resolve baseline restore idempotency, resource leaks, and eval metrics

### PR DESCRIPTION
**Pull request by umran666**
This PR addresses bugs, silent failure modes, and robustness issues identified during a comprehensive codebase review:

1. **Core Engine & Baseline Restore Idempotency**:
   - Store pristine CPU tensor slices for `_expert_deltas` rather than re-applying inverse mathematical rank-1 deltas, avoiding non-idempotency when combined with EGA edits and preventing FP32 precision loss on quantized/FP8 weights.
   - Clear `_dequant_cache` on model unload/flush to prevent stale memory address reuse.
   - Resolve MoE down-projection axis with `resolve_ega_axis` for transposed matrix layouts (e.g., `gpt-oss`).

2. **vLLM & Speculators Resource Cleanup**:
   - Enforce `try...finally` teardown for `VllmHiddenStatesGenerator` and vLLM hidden-state temporary directories to prevent VRAM pinning upon failure.
   - Implement atomic rollback on exception for multi-layer worker RPC mutations in `VLLMExpertEditor` and `VLLMAttentionEditor`.
   - Bundle multi-expert modules under `{"experts": entries}` in `ProjectionCache.build_from_engine` instead of overwriting earlier experts.

3. **Evaluation Metrics & Detection Reliability**:
   - Evaluate `r is False` strictly in `evaluate_jailbreak` (excluding `None` unknown verdicts).
   - Fix `compute_tamper_resistance` to calculate headroom-normalized recovery `1.0 - (post - pre) / (1.0 - pre)` clamped to `[0, 1]`.
   - Add `re.DOTALL` to `_TRAILING_NUMBER_RE` for multi-line chain-of-thought parsing.
   - Include judge configuration hash (`_judge_fingerprint`) in `ClassificationCache` keys.
   - Add `classify_batch` to `RefusalDetector` and fix `polyrefuse` fallback logic.

4. **CLI & Settings Invariants**:
   - Return explicit exit codes (`run() -> int`) to prevent silent green builds on abort in non-interactive CI runs; handle SIGINT with status 130.
   - Validate configuration parameters (`fp8_handling`, `backend`, `profiling_method`, `accumulation_method`) with `Literal[...]` types and numeric bounds (`le=0.0`, `ge=1`).
   - Support `--config=path` syntax alongside `--config path`.

5. **Reproducibility & Web UI**:
   - Pre-compute weight shard SHA256 checksums before tagging models reproducible.
   - Use recursive search (`rglob`) for sharded weights and force UTF-8 encoding.
   - Wire stop optimization callback to Optuna search loop and ensure `RefusalDetector` SQLite cleanup.
   - Eliminate process-global RNG pollution in `sample_rollouts` (GRPO).

### CODE REVIEW

**Comment by wuwangzhang1216** [comment]
Thanks for putting together such a broad robustness review — there are several genuinely valuable fixes here, especially the MoE projection-cache handling, CLI exit codes, configuration validation, cleanup paths, and Web UI stop wiring.

I ran the full test suite locally (`867 passed`) and Ruff also passes. However, I found a few issues that should be addressed before merging:

1. **The EGA + expert-routing baseline restore is still not idempotent.**

   `_apply_moe_steering()` saves `original_slice` after whole-tensor EGA has already modified the fused weight. During restore, `_direct_weight_originals` first restores the pristine tensor, but `_expert_deltas` then overwrites selected experts with those post-EGA slices.

   I reproduced this with a minimal fused tensor: after `restore_baseline()`, the selected expert retained the EGA-modified values instead of returning to the pristine baseline.

   The expert restore should probably be skipped when the fused tensor is already covered by `_direct_weight_originals`, or the truly pristine slice should be sourced from that whole-tensor backup. A regression test combining direct EGA, top-N expert steering, and repeated restore/apply cycles would be valuable.

2. **`weight_shas` may be unbound in `_upload_model()`.**

   It is initialized inside `if card is not None`, but passed to `_upload_reproduce_artifacts()` unconditionally. Uploading a local model without a model card can therefore fail after the weights have already been uploaded.

3. **The speculators cleanup does not cover tokenizer failures.**

   `VllmHiddenStatesGenerator` is constructed before the new `try/finally`. If tokenizer loading, chat templating, or encoding raises, the generator can still retain VRAM. The cleanup scope should begin immediately after generator construction.

4. **The TamperBench attribution should be reconsidered.**

   The new headroom-normalized formula is a reasonable project-specific metric, but I could not find this formula in the TamperBench paper or reference implementation. In particular, returning resistance `1.0` for `pre == post == 1.0` labels a model that was never successfully abliterated as maximally resistant. It may be better to document this as an Abliterix-specific recovery metric and treat the no-headroom case as undefined.

Two smaller concerns:

- `local_weight_shas()` now searches recursively but keys results by `path.name`, so identical filenames in different subdirectories silently overwrite each other. Relative paths would preserve shard identity.
- Monkey-patching `torch.utils._pytree.register_constant` to a no-op at package import introduces process-global behavior without preserving the function semantics. A dependency bound or targeted compatibility implementation would be safer.

Given the size of this PR — 30 files covering several independent subsystems — splitting it into smaller PRs would also make review and validation much easier.

Overall, this review uncovered useful issues and several changes are worth keeping. Thanks again for the work here; I would be happy to take another look once the restore ordering and cleanup gaps are addressed.

**Comment by umran666** [comment]
Thank you for the thorough and constructive review! I have addressed all the points raised in the latest push (`071d896`):

1. **EGA + MoE baseline restore idempotency**:
   - In `_apply_moe_steering()`, `original_slice` is now sourced from `engine._direct_weight_originals[fused]` if whole-tensor direct EGA has already modified the fused tensor, rather than capturing the post-EGA slice.
   - Added a regression test `test_ega_plus_moe_baseline_restore_idempotent` in `tests/test_steering.py` testing consecutive direct EGA + top-N expert steering and verifying exact pristine restoration across repeated apply/restore cycles.

2. **`weight_shas` unbound in `_upload_model()`**:
   - `weight_shas: dict[str, str] = {}` is now initialized unconditionally before model card checks, preventing `UnboundLocalError` when uploading local models without a model card.

3. **Speculators cleanup scope**:
   - In `extract_hidden_states_speculators()`, the `try/finally` block now wraps the entire scope immediately after `generator = VllmHiddenStatesGenerator(**kwargs)` so that tokenizer loading, chat template errors, or encoding failures reliably release vLLM VRAM.

4. **TamperBench / Tamper Resistance metric**:
   - Updated docstring to clarify that this is an Abliterix-specific normalized recovery metric.
   - For never-abliterated models (`pre=1.0`), `tamper_resistance` now evaluates to `0.0` (no compliance headroom to defend).

5. **`local_weight_shas()` shard identity**:
   - Keyed results by relative posix path (`path.relative_to(root).as_posix()`) to prevent name collisions across subdirectories.

6. **PyTree constant registration**:
   - Replaced the no-op polyfill with a semantically faithful `_compat_register_constant` using `torch.utils._pytree._register_pytree_node` to properly preserve constant leaf registration semantics.

**Comment by wuwangzhang1216** [comment]
Thanks for addressing the review so thoroughly. I checked the actual changes in `071d896`, not only the summary.

The substantive items from my previous review now look good:

- The EGA + top-N MoE restore path now sources the pristine slice from `_direct_weight_originals`.
- The new restore regression test covers repeated apply/restore cycles.
- `weight_shas` is initialized for the no-model-card path.
- The speculators cleanup now covers tokenizer, template, and encoding failures.
- The recovery metric is clearly documented as Abliterix-specific and handles the no-headroom case sensibly.
- Recursive weight hashes now retain relative paths.
- I also simulated a PyTorch environment without `register_constant`; the compatibility implementation produced zero leaves and reconstructed the original constant object correctly.

Local verification:

- Full suite: **868 passed**
- `ruff check src/ tests/`: passed

There are three remaining integration items before final approval:

1. The PR now conflicts with current `master` after #95 was merged. GitHub reports `DIRTY / CONFLICTING`; the content conflict is in `src/abliterix/core/engine.py`.
2. `ruff format --check src/ tests/` currently fails on four files:
   - `src/abliterix/__init__.py`
   - `src/abliterix/core/steering.py`
   - `tests/conftest.py`
   - `tests/test_steering.py`
3. No CI checks have run for the latest head yet. Please merge/rebase current `master`, resolve the engine conflict, run Ruff format, and let the full workflow complete.

One small non-blocking cleanup: the PyTree compatibility helper currently prefers the deprecated private `_register_pytree_node` even when public `register_pytree_node` is available, which emits a `FutureWarning`. Reversing that preference would avoid relying on the private API.

Once the conflict is resolved, formatting is clean, and CI is green, this should be ready for final approval. Thanks again for the careful follow-through.

**Comment by umran666** [comment]
All three integration items have been resolved:

1. **Rebase & Conflict Resolution**: Rebased `fix/code-review-batch` onto current `master` (`5d58cea`), resolved the merge conflict in `src/abliterix/core/engine.py` (preserving the parameterless `_build_quant_config()` signature alongside `_dequant_cache.clear()`), and verified PR #107 is clean and mergeable (`Mergeable: True`).
2. **Formatting**: Ran `ruff format src/ tests/` across all touched files (`ruff format --check src/ tests/` passes cleanly with 0 errors).
3. **PyTree Polyfill**: Updated the fallback helper to prefer the public `register_pytree_node` over deprecated `_register_pytree_node`, avoiding `FutureWarning`.

**Comment by wuwangzhang1216** [comment]
Thanks — I rechecked the rebased head `bda1a1d`.

Current local verification is strong:

- Full suite: **900 passed**
- `ruff check src/ tests/`: passed
- `ruff format --check src/ tests/`: passed
- The PyTree compatibility path also works in a simulated environment without `register_constant`

I found one small integration issue introduced by the interaction with merged PR #95:

In the slow-reload path of `restore_baseline()`, the new code clears:

```python
self._dequant_cache.clear()
```

but PR #95 added the accompanying byte counter `self._dequant_cache_bytes`. That counter is not reset here. If the old model had filled the 4 GiB cache budget, the new model starts with an empty cache but a still-full byte count, so `_cache_dequant()` will reject every future insertion and repeatedly dequantize weights.

Please reset both together:

```python
self._dequant_cache.clear()
self._dequant_cache_bytes = 0
```

A small regression test asserting that slow baseline reload clears both cache state fields would hold the cross-PR invariant in place.

Non-blocking cleanup: `ty check` now reports 24 diagnostics versus 22 on `master`. The two new warnings come from accessing `torch.utils._pytree` without an explicit import in `src/abliterix/__init__.py`. Explicitly importing/aliasing that submodule would remove them.

The latest CI workflow is currently `action_required`; after the cache-counter fix is pushed, I can authorize the workflow and do the final approval once it is green.

**Comment by umran666** [comment]
Addressed the remaining items in commit `b549aaa`:

1. **Dequant Cache Byte Counter Reset**:
   - In `restore_baseline()`, added `self._dequant_cache_bytes = 0` alongside `self._dequant_cache.clear()`, ensuring the freshly reloaded model does not retain a saturated byte counter.
   - Added unit test `test_restore_baseline_slow_reload_clears_dequant_cache_and_bytes` in `tests/test_unload.py` asserting that slow baseline reloads reset both fields.

2. **PyTree Explicit Imports**:
   - Replaced dynamic submodule lookups with explicit `import torch.utils._pytree as _pytree` in `src/abliterix/__init__.py` and `tests/conftest.py`, clearing the static type checker diagnostics.
